### PR TITLE
py/modmath: Add full checks for math domain errors.

### PR DESCRIPTION
**Pull request by dpgeorge**
This patch adds full checking in the `math` module for mathematical domain errors, like `acos(10)`.  Python is a relatively safe language because it has good error checking and reporting, and having checks on math functions helps to write safe and correct numerical code.

But of course it costs in terms of code size (and some execution overhead).  The code size increase due to this patch is (for ports that have the `math` module enabled):
```
   unix x64:  +224
unix nanbox:  +304
      stm32:  +216
    esp8266:  +268
```
That's a fair but, but if a port has floating point then it's already paying much more this to get float support at all.

It's easy to add a config option to control whether domain checks are made in the math module (eg MICROPY_PY_MATH_DOMAIN_CHECKS).  It's also possible to have some checks always enabled (like the existing check for square root of a negative number) and others controlled by a config option, but IMO that just gets too messy.

In summary: I'd suggest that either 1) the checks are always enabled, or 2) all of them are configured by a single config option (best enabled by default so it's safe by default).

### CODE REVIEW

**Comment by pfalcon** [comment]
So, if you mean like moving existing checks in the module to use the same "enable" config option as newly introduce checks, then yes, makes sense. I'm less sure about enabling such an option by default. We can enable it for Unix port, I may imagine that you'll enable it for stmhal, but other ports by default will probably want it off.

**Comment by dpgeorge** [comment]
> So, if you mean like moving existing checks in the module to use the same "enable" config option as newly introduce checks

Yes that's what I meant.  The existing checks are in sqrt, log, log2 and log10.  Although now looking at it again I don't think it's a good idea to disable the sqrt and log checks because they are pretty fundamental checks.  And pow, asin and acos checks fall in the same class, that's why I think the checks should be enabled by default.

**Comment by pfalcon** [comment]
> they are pretty fundamental checks. And pow, asin and acos checks fall in the same class, that's why I think the checks should be enabled by default.

Used carelessly, this approach ("need to add more stuff") can be applied to almost anything. On the other hand, we somehow lived 3+ years without it. Or let's take an (in-tree) example like tests/misc/rge_sm.py. Would it stop working without those checks?

Because another approach is: "if a function expects some range of data, it's fair to expect that that range will be used". And for developing or debugging the code, there's CPython. And this approach was recommended for a long time.

Anyway, with all the above, if you really think that this place (out of hundreds other) is what requires extra checks on by default, well, let's do it, as long as there's a way to disable those checks consistently. Please add Zephyr port to the usual set of ports where it's disabled.

**Comment by stinos** [comment]
+1 these are pretty fundamental checks. I consider acos(10) and e.g. 1/0 the same: returning a value here, whatever it is, is simply incorrect, this is math.

**Comment by dpgeorge** [comment]
See #3347 for an improved version of this PR.

**Comment by dpgeorge** [comment]
The alternative PR to this was merged, see 25e140652b946e9a136a16a9831c0b18f1285702
